### ISSUE TITLE
Change Request: Adding an additional format for the UA string

### ISSUE THREAD

**Issue by mayurid** [Azure/azure-sdk#1457]
## The Basics

* Which design guideline is affected?
General Guidelines -  Telemetry policy (https://azure.github.io/azure-sdk/general_azurecore.html#telemetry-policy)

* Which target languages are affected?
All languages

* Describe the change:
The Cosmos team would like to introduce an alternate blessed UA string format:
azsdk-<sdk_language>-<package_name>/<package_version> <platform_info> [(Feature Flags)] [<application_id>]
            <application_id>: suffix instead of prefix
            (Feature Flags): Encoding to represent the client options, its useful for telemetry & 
            troubleshooting purpose. 

## Reasoning
Per @kirankumarkolli : With current Cosmos monitoring infrastructure its not practical to prefix the application id. 
Our monitoring indexes the first parts of the user-agent and expects it to be a fixed set of values. 
With variables names the dimension cardinality is unbounded and resulted in Kusto service issues.

**Comment by kyle-patterson** [comment]
A little more info:

The [current guidelines](https://azure.github.io/azure-sdk/general_azurecore.html#telemetry-policy) expect the form:

```
[<application_id> ]azsdk-<sdk_language>-<package_name>/<package_version> <platform_info>
```

Due to a server-side restriction, Cosmos cannot support an unbounded-length [<application_id> ] prefix. The end result is a possibly-truncated UserAgent string. There are two proposed approaches so far:

### PostFix User Agent Notation
Swap the [ <application_id>] to be a suffix and accept the possible truncation. This could impact our metrics system, which relies on the application_id field to identify libraries such as Spring Cloud.
```
azsdk-<sdk_language>-<package_name>/<package_version> <platform_info>[ <application_id>]
```

### Bounded-Length User Agent
Restrict the [<application_id> ] to be within a bounded length. This prevents truncation, but limits the availability of the application_id and any feature flags.

## Client Hints
Another possibility is to break the user agent string into smaller fields, similar to [Client hints](https://wicg.github.io/ua-client-hints/#examples). In this approach, x-ms-useragent could be frozen but not deprecated:

```
UserAgent: [<application_id> ]azsdk-<sdk_language>-<package_name>/<package_version> <platform_info>
X-MS-AZSDK-Telemetry: class=BlobClient;method=DownloadFile;blobType=Block

x-ms-azsdk-appid: [<application_id>]
x-ms-azsdk-pkg: azsdk-<sdk_language>-<package_name>/<package_version>
x-ms-azsdk-platform: <platform_info>
x-ms-azsdk-method: class=<class_name>;method=<method_name>;<Key-Value_pairs>
```

**Comment by ellismg** [comment]
The existing guidelines say the following about `application_id`:

> DO enforce that the application ID is no more than 24 characters in length. Shorter application IDs allows service teams to include diagnostic information in the “platform information” section of the user agent, while still allowing the consumer to obtain telemetry information for their own application.

Which to me sounds like we do not expect to have application ID without unbounded length.

My reading of the initial issue makes it seem like the problem is not that application id is unbounded in length, but rather there are not a fixed set of values, which makes it difficult for some back-end services to ingest into their monitoring pipelines, because they require a fixed set of prefixes for user-agent strings and since application id can be set to an arbitrary value (even with an upper bound on length) it breaks this invariant.

**Comment by JeffreyRichter** [comment]
I don't think it's a big deal to support 1 format or another. I think the problem now is that the ship has sailed. We have SDKs that use the documented format and these SDKs will live in the wild for many years to come. If we adopt a "new" format then we are really adopting "another" format. Our internal tools will have to support both of these formats and unify the results. This is where the bulk of the work/effort will be if we adopt another format. I don't have a good sense of the effort required here. Changing the format in the new SDK versions is trivial.

**Comment by johanste** [comment]
The proposed change violates the convention of user agents where the entries are listed in decreasing order of significance (see https://tools.ietf.org/html/rfc7231#section-5.5.3) - the application is presumably the *most* significant part of the user agent, and thus should be first. I'd have to better understand what the underlying problem is here...

**Comment by lilyjma** [comment]
[MS only] [video](https://msit.microsoftstream.com/video/891ba4ff-0400-a521-5d19-f1eacac28b36)

**Comment by kushagraThapar** [comment]
This is done and will be released in 4.3.1
